Applications/Spire/Source/KeyBindings/TaskKeysPage.cpp

```diff
@@ -42,16 +44,219 @@ namespace {
   }
 }
 
+struct OrderTaskMatchCache {
+  std::unordered_set<QString> m_caches;
+
+  bool matches(const OrderTaskArguments& order_task, const QString& query) {
+    if(m_caches.contains(query)) {
+      return true;
+    }
+    auto matched = ::matches(order_task, query);
+    if(matched) {
+      m_caches.insert(query);
+    }
+    return matched;
+  }
+};
+
+struct TaskKeysPage::FilteredTaskKeysListModel : OrderTaskArgumentsListModel {
+  std::shared_ptr<OrderTaskArgumentsListModel> m_source;
+  std::vector<OrderTaskMatchCache> m_caches;
+  std::vector<int> m_filtered_data;
+  QString m_query;
+  ListModelTransactionLog<OrderTaskArguments> m_transaction;
+  scoped_connection m_source_connection;
+
+  explicit FilteredTaskKeysListModel(
```

Comment:
This looks significantly more complicated than using the existing ```FilteredTableModel```, can you give some reasons why this was used instead?

Reply:
The model is a "ListModel". Since `ListModel` lacks a filter model, I implemented this `FilteredModel` based on the implementation of `FilteredTableModel`. If using an existing `FilteredTableModel`, the approach I can think of is to convert the `ListModel` to a `TableModel` for filtering. Because "TaskKeysTableView" has already performed a conversion from "ListModel" to "TableModel", I think converting again would be redundant and inefficient. Do you have any suggestions for using the FilteredTableModel?

Reply:
Convert from a ```ListModel``` to a ```TableModel``` and then use a ```FilterTableModel``` on it.

Reply:
Can I update the first parameter of `make_task_keys_table_view` from `std::shared_ptr<OrderTaskArgumentsListModel>` to `std::shared_ptr<TableModel>` to reduce the need for conversion from `ListModel` to `TableModel`? Otherwise, `ListModel` would have to be converted to `TableModel` to use `FilterTableModel`, and the converted `TableModel` would be converted back to `ListModel` before being passed to `make_task_keys_table_view`.

Reply:
Yes there's no issue with that.